text/000-rustdoc-jump-to-definition.md

```diff
@@ -0,0 +1,187 @@
+
+ - Use Ctrl-F to search the page for either a definition of `OtherStruct` or a `use` statement that imports it.
+ - If you found a `use` statement, figure out which file it maps to.
+ - Open the sidebar and navigate to that file.
+ - Ctrl-F again to find the actual definition of `OtherStruct`.
+ 
+If `OtherStruct` itself is defined in terms of additional structs, you may need to repeat this process many times to get even a cursory understanding of the implementation of `MyStruct`. It can also be error prone. If we implement "go to definition", we can turn this whole process into a single click, and make the result reliable.
+
+The same applies when reading a source code when type annotations are not available:
+
+```
+let var = some_other_var.do_something();
+let var2 = a_function();
+```
+
+It would require to look for each method in the various files, etc.
+
+So overall, this would greatly improve the source code browsing experience.
+
+The other goal is to allow to go back from the source code pages directly to the item documentation to be able to read the rendered content and not just the "raw" doc comments.
+
+So in short, this feature is split in two parts:
+
+ * Jump to the definition of an ident.
+ * Jump to an item documentation from the source view.
+
+Adding this information directly into the documentation allows it to improve the documentation value: if you want to go a bit further what is written in the documentation (to see some implementation details for example), you can currently go to the source code. However, you're quickly limited to the current file if it's using private items not from this file.
+
+Once you found what you're looking for, having a link to an item's documentation from the source code allows to add the missing connection between the source code viewer and the doc pages.
+
+Another important note is that if documentation is split in multiple places or requires an external tool for the source code navigation, it's actually slowing down our users quite a lot.
+
+A good example where this would be very helpful is for the [rust compiler source code](https://doc.rust-lang.org/1.57.0/nightly-rustc/src/rustc_middle/hir/map/mod.rs.html#872-877): even if a fully set up IDE or github, it's very hard to go around without previous knowledge or help from someone else. It's also very common to have undocumented items or to just want to see how something is implemented. With the "jump to definition", it already improved quite a lot the browsing experience by allowing to jump super quickly by simply clicking on a item. The missing part is now to go back from the source code into the item documentation to make it complete.
+
+# Reference-level explanation
```

Comment:
This section really needs some work.

At a minimum:

A list of exactly what kinds of links are happening. Type names to types, function names to functions, trait names to traits, macro names to macros, probably. const names to consts and static names to statics? Should generic types have each individual part of the generic type link correctly?

Are we also doing variable names to variable types or their local variable definitions? If we're linking them to their types (which is what it _seems_ like from this RFC), how about variables involving generics? And should we special case smart pointers (e.g. should a variable with type `&Foo` link to `Foo`? What about `Box<Foo>`? Where do we draw the line?)

And what about doc examples? It would seem that JTD in doc examples would be far more useful than in source, but also far trickier to implement especially since _currently_ `rustdoc` does not compile examples except when running tests.

This is kinda what I mean by pin down the scope, this RFC is _still_ rather broad and underspecified, there's a lot of work that needs to be done to pin down stuff like the above. The above is what I could think of quickly whilst writing this comment, we need to spend time brainstorming to figure out all potential edges of this figure.

We should have a subsection of this section that talks about "implementation details" too, containing the stuff you have below. But first, we ought to pin down the design.